### PR TITLE
Add support for Elixir

### PR DESCRIPTION
**Pull request by stevedomin**
From @greysteil's PR https://github.com/gocardless/bump-core/pull/57
> A starting point for adding Elixir support.
> 
> Next step would be to write a parser for the mix.lock file, in Elixir. The JavaScript helper run task and parser are the best places to start for understanding how shelling out to helpers works. We'll probably want to piggy back off of Hex as much as possible here, and include it as a dependency (like we do with Yarn).
> 
> Once an Elixir parser is written, the pending specs in spec/bump/file_parsers/elixir/hex_spec.rb should pass. At that point, let's talk about adding the UpdateChecker, FileUpdater and MetadataFinder classes.

### CODE REVIEW

**Comment by greysteil** [comment]
Update: @stevedomin and I are going to make this happen on Wednesday 18th. 🎉

**Comment by greysteil** [comment]
@stevedomin - what's the best next step on this? Don't want it to lose momentum. Can you push whatever you've got on your machine and add a TODO list of what's remaining?

**Comment by stevedomin** [comment]
@greysteil please before pulling this branch again can you make sure you checkout your local branch to a temp branch, I accidentally removed your commits around metadata.

**Comment by stevedomin** [comment]
Thanks @greysteil! :)

**Comment by stevedomin** [comment]
@greysteil got the `UpdateChecker` close to finished. Once this is done there is only the `FileUpdater` left, right?

**Comment by greysteil** [comment]
That's awesome! And yep, that sounds right to me. Want me to take a look through what's here now?

File: helpers/elixir/bin/run.exs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+          "MIX_LOCK" => nil,
+          "MIX_DEPS" => nil
+        }
+      ]
+    )
+    :erlang.binary_to_term(output)
+  end
+end
+
+# %Mix.Dep{
```
Comment (stevedomin) [inline helpers/elixir/bin/run.exs]: This is going to go away before we merge into master

**Comment by stevedomin** [comment]
Yep, go for it!

**Comment by greysteil** [comment]
I'm going to need to do a bit rebase on this. Can you let me know when everything you've got is pushed up, and I'll get it done?

**Comment by greysteil** [comment]
Rebase done.

**Comment by greysteil** [comment]
I *think* I've got the specs passing. 🎉

**Comment by greysteil** [comment]
^^^ It worked!!

OK, remaining TODOs:
- [ ] A requirements updater (for the update checker)
- [ ] More specs for the parser (e.g., what does it do when no versions is specified? What does it do for git dependencies)
- [ ] A FileUpdater

@stevedomin - are you OK to do the `FileUpdater`? I'll do the requirement updater (it can be in Ruby).

Also, what's happening under the hood in the UpdateChecker? Is Dependabot doing a full install, or just piggy-backing off of mix's resolution logic. Because if it's the former it could be really slow...

**Comment by stevedomin** [comment]
Awesome @greysteil, thank you so much!

Right now, it's doing a full install but I'm working on a version without it. I wanted to get something out of the door, been trying to get it working without the install for way too long.

We can definitely make it work without doing a full install but I'll need to make some changes to the `Mix` codebase itself though.

Are you ok with getting a working version like this and then once it's out we can figure out to do this without the full install?

I'll do the FileUpdater and add specs to the parser.

**Comment by greysteil** [comment]
Makes sense to me - if you get the FileYpdater done I’m 👍 on shipping

**Comment by stevedomin** [comment]
Awesome!

**Comment by greysteil** [comment]
`RequirementsUpdater` is done, and pretty solid.

One question - for equality, is `"== 1.0.0"` a valid specifier? And if so, is `"= 1.0.0"` valid or invalid? We're going to need to handle that if so...

**Comment by greysteil** [comment]
@stevedomin - I've added a `FileUpdater`, and done the mixfile updating logic. Just needs your Elixir code to do the lockfile updating and this will be basically good to do.

**Comment by greysteil** [comment]
@stevedomin - added a couple of failing specs for you. Only thing needed now is the Elixir code to update the lockfile, which I *think* you can basically pinch from what you wrote for the `UpdateChecker`...

**Comment by greysteil** [comment]
@stevedomin - I think this is all working, so I'm going to merge and release. Would love your help improving and testing it, though!

**Comment by stevedomin** [comment]
Awesome work and thanks @greysteil. I'll add it to my repos tomorrow.

And I'll check the codebase to see what we can improve. At the very least I'd like to change UpdateChecker so that it doesn't pull the dependencies.

**Comment by greysteil** [comment]
That would be a huge improvement! 🎉

**Comment by greysteil** [comment]
@stevedomin could I get an IP assignment comment from you, just so it's explicit what your intentions were with this PR? Ideally the following text would be amazing:

> I'm gifting the IP of the contents of this PR and all previous and future contributions to dependabot-core to Dependabot for it to be used however Dependabot wishes.

Thanks. ❤️

**Comment by stevedomin** [comment]
I'm gifting the IP of the contents of this PR and all previous and future contributions to dependabot-core to Dependabot for it to be used however Dependabot wishes.

**Pull request merged**